### PR TITLE
Add support for Soy Element Composition.

### PR DESCRIPTION
**Pull request by iteriani**
Add support for Soy Element Composition. It has the syntax in the form of

<{foo()}></>

This adds support to pass through allowEmptyTag and to support this mode in closetag.

### CODE REVIEW

File: mode/xml/xml.js [hunk 1]
```diff
@@ -399,7 +399,7 @@ CodeMirror.defineMode("xml", function(editorConf, config_) {
     xmlCurrentContext: function(state) {
       var context = []
       for (var cx = state.context; cx; cx = cx.prev)
-        if (cx.tagName) context.push(cx.tagName)
+        context.push(cx.tagName)
```
Comment (marijnh) [inline mode/xml/xml.js:402]: This changes the return value of this function, which looks like it will confuse the closetag and xml-hint addons. Why is this change needed?

File: mode/htmlmixed/htmlmixed.js [hunk 2]
```diff
@@ -74,7 +74,8 @@
       name: "xml",
       htmlMode: true,
       multilineTagIndentFactor: parserConfig.multilineTagIndentFactor,
-      multilineTagIndentPastTag: parserConfig.multilineTagIndentPastTag
+      multilineTagIndentPastTag: parserConfig.multilineTagIndentPastTag,
+      allowMissingTagName: parserConfig.allowMissingTagName,
```
Comment (marijnh) [inline mode/htmlmixed/htmlmixed.js:78]: It doesn't look like this is used at all.

Comment (iteriani) [reply mode/htmlmixed/htmlmixed.js:78]: This is passed through to xml (https://github.com/codemirror/CodeMirror/blob/master/mode/xml/xml.js#L55)

File: mode/xml/xml.js (hunk 1 above)
Comment (iteriani) [reply mode/xml/xml.js:402]: Right now what happens when you do something like

<{foo()}> or even <>, the bit here

https://github.com/codemirror/CodeMirror/blob/master/addon/edit/closetag.js#L131

causes context[context.length - 1] to be context[-1], which is undefined. So the autocomplete would have

<{foo()}></undefined>

File: mode/htmlmixed/htmlmixed.js (hunk 2 above)
Comment (marijnh) [reply mode/htmlmixed/htmlmixed.js:78]: Ah, right, my bad.

**Comment by marijnh** [comment]
Thanks for clarifying.

**Comment by iteriani** [comment]
Thanks a lot! This was a great experience.
